src/main/kotlin/edu/carroll/cs389/service/tracker/TrackServiceImpl.kt

```diff
@@ -0,0 +1,113 @@
+     *
+     * @return A boolean representing success in saving the information to the database
+     */
+    override fun trackClient(
+        ipv4: String?,
+        browser: String?,
+        browserMajorVersion: String?,
+        platform: String?,
+        platformVersion: String?,
+        uri: String?): Boolean {
+        // if we are missing and ipv4 or uri then abort
+        if (ipv4 == null || ipv4 == "Unknown" || ipv4 == "") {
+            log.error("trackClient: client does not have a valid connecting ipv4 address, request invalid, aborting")
+            return false
+        }
+        if (uri == null || uri == "Unknown" || uri == "") {
+            log.error("trackClient: client is attempting to connect to a page that does not exist, request invalid, aborting")
+            return false
+        }
+
+        // if we have a browser and corresponding version
+        var saveBrowser: String? = null
+        if (browser != null && browser != "" && browser != "Unknown" &&
+                (browserMajorVersion != null && browserMajorVersion != "Unknown" && browserMajorVersion != "")) {
+            log.debug("trackClient: $ipv4 successfully acquired browser info")
+            saveBrowser = "$browser $browserMajorVersion"
+        }
+        // if we have a browser but not a corresponding version
+        else if (browser != null && browser != "" && browser != "Unknown") {
+            log.warn("trackClient: $ipv4 client browser missing version tag")
+            saveBrowser = browser
+        } else {
+            log.warn("trackClient: $ipv4 client has no browser info")
+        }
+
+        // null-check and format operating system string (java library has potential to return null)
+        var os: String? = null
+        // if we have a platform and corresponding version
+        if (platform != null && platform != "" && platform != "Unknown" &&
+            (platformVersion != null && platformVersion != "Unknown" && platformVersion != "")) {
+            log.debug("trackClient: $ipv4 successfully acquired OS info")
+            os = "$platform $platformVersion"
+        }
+        // if we have a platform but not a corresponding version
+        else if (platform != null && platform != "" && platform != "Unknown") {
+            log.warn("trackClient: $ipv4 client OS missing version tag")
+            os = platform
+        } else {
+            log.warn("trackClient: $ipv4 client has no OS info")
+        }
+
+        // save trackedUser to DB with potentially null val for OS/Browser
+        val user: TrackedUser = TrackedUser(ipv4, os, saveBrowser, uri)
+        log.debug("trackClient: Successfully created TrackedUser entity for client $ipv4")
+        return try {
+            trackerRepo.save(user)
+            log.info("trackClient: Successfully tracked $ipv4")
+            true
+        } catch (e: Exception) {
+            log.error("trackClient: Failed to save client info to database")
```

Comment:
log.error("trackClient: Failed to save client info to database", e)

Also, don't catch 'Exception', but the error you are 'expecting' (??) to be thrown.